apollo-router/src/uplink/license_enforcement.rs

```diff
@@ -591,28 +596,57 @@ impl Display for LicenseEnforcementReport {
     }
 }
 
-/// License controls availability of certain features of the Router. It must be constructed from a base64 encoded JWT
+/// These are claims extracted from the License and represent ways Apollo limits the router's usage It must be constructed from a base64 encoded JWT
 /// This API experimental and is subject to change outside of semver.
 #[derive(Debug, Clone, Default)]
 pub struct License {
     pub(crate) claims: Option<Claims>,
 }
 
+/// Transactions Per Second limits. We talk as though this will be in seconds, but the Duration
+/// here is actually given to us in milliseconds via the License's JWT's claims
+#[derive(Builder, Copy, Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub(crate) struct TpsLimit {
+    pub(crate) capacity: usize,
+    pub(crate) interval: Duration,
+}
+
+/// LicenseLimits represent what can be done with a router based on the claims in the License. You
+/// might have a certain tier be limited in its capacity for transactions over a certain duration,
+/// as an example
+#[derive(Debug, Builder, Copy, Clone, Default, Eq, PartialEq)]
+pub(crate) struct LicenseLimits {
+    /// Transaction Per Second limits. If none are found in the License's claims, there are no
+    /// limits to apply
+    pub(crate) tps: Option<TpsLimit>,
+}
+
 /// Licenses are converted into a stream of license states by the expander
 #[derive(Debug, Copy, Clone, Eq, PartialEq, Default, Display)]
 pub(crate) enum LicenseState {
     /// licensed
-    Licensed,
+    Licensed(Option<LicenseLimits>),
```

Comment:
Should this be `LicenseLimits` without the `Option` here, since `tps` is `Option`? Basically, we'd have 2 layers of optionality without semantic advantage.

Reply:
agreed; will remove it!

Reply:
changed around a bit, but decided to keep it; no license limits is differen than a license with limits but no tps limits, so both options are needed